Allow skin elements to find closest anchor

ribbanya opened the pull request:
Resolves ppy/osu#13252.

There are no tests. I chose not to implement tests because `SkinSelectionHandler` does not have any yet, and all of the logic is private to that class. If tests are requested, I will try my best to implement them.

I also implemented localisation for these menus, because the menu item text needed to be decoupled from `Anchor` values anyway.

In `SkinSelectionHandler`, I basically just threw the new code at the top of the file, because I wasn't sure what the idomatic ordering of members is. I'd be happy to rearrange it if no one else wants to.

https://user-images.githubusercontent.com/5190964/120913755-6b5b2a80-c667-11eb-976f-693eb3e796b7.mp4

ribbanya committed:
Allow skin elements to find closest anchor

- Resolves ppy/osu#13252
- Add localisation strings for the context menu instead of using enum

gagahpangeran commented:
Hello, thank you for your contribution.

For future reference, please avoid only a single commit for large/non-trivial changes like this PR.

frenzibyte commented:
It would've been a better choice to discuss how this should be implemented rather than putting the effort to work on it with the way you have it.

I would've gone with a private `usingClosestAnchor` bindable or a `SkinAnchor` enum (depends on whether `Origin` should have "Closest" as well) rather than all the hash logic you have here. Would be much simpler and easier to understand.

Also, I'm not really sure about including localisation logic like that yet, that's out of scope from this PR in general and is adding more complexity to review this. The smaller the PR changes are, the easier it is to review and merge.

Also, a test is definitely required for this, if you're adding new functionality to the game, you should better have a test scene for it that would fail if it's not working properly, and it should be quite feasible to write one for such functionality. (just need to add a `SkinSelectionHandler` with a random skin drawable inside, and start moving them using a `ManualInputManagerTestScene` and do whatever you should for testing)

frenzibyte reviewed (changes requested):
As pointed out.

bdach commented:
> depends on whether `Origin` should have "Closest" as well

What would that even mean? The origin would be closest relative to what?

> I'm not really sure about including localisation logic like that yet, that's out of scope from this PR in general and is adding more complexity

I think the bigger thing is that I'm not sure we have all the parts of the localisation system worked out for it to be 100% ready for consumption.

ribbanya commented:
> Hello, thank you for your contribution.
> 
> For future reference, please avoid only a single commit for large/non-trivial changes like this PR.

Yeah, that was my bad. Sorry.

> It would've been a better choice to discuss how this should be implemented rather than putting the effort to work on it in the way you have it.

Hm, that's fair. I guess I should join the discord? I did comment in #13252 with two proposed methodologies yesterday, though.

> I would've gone with a private `usingClosestAnchor` bindable

In what class? Would that be a `BindableBool`?

> Also, I'm not really sure about including localisation logic like that yet, that's out of scope from this PR in general and is adding more complexity to review this. The smaller the PR changes are, the easier it is to review and merge.

Okay. Should I split that off into a new PR? I could also just revert back to using strings from the enum and forget about localisation.

> Also, a test is definitely required for this

Okay, I'll have to study how your testing works first.

@frenzibyte Based on the scale of your requested changes, should I close/draft this PR and come back with a new one?

frenzibyte commented:
> > depends on whether `Origin` should have "Closest" as well
> 
> What would that even mean? The origin would be closest relative to what?

I just put that out as I'm not sure myself, not asking for it to be a thing. It's not clear in the issue thread whether it should be a thing either, it's just stated there alongside anchor.

frenzibyte commented:
> > I would've gone with a private `usingClosestAnchor` bindable
> 
> In what class? Would that be a `BindableBool`?

in `SkinSelectionHandler` itself, and it would be updated whenever any of the menu items are clicked. Though now that I think about it, could just go with the latter proposal anyway (`SkinAnchor` enum that has a `Closest` as well) and use that for the Anchor menu item, would be simpler than the bindables way.

> > Also, I'm not really sure about including localisation logic like that yet, that's out of scope from this PR in general and is adding more complexity to review this. The smaller the PR changes are, the easier it is to review and merge.
> 
> Okay. Should I split that off into a new PR? I could also just revert back to using strings from the enum and forget about localisation.

Probably, but also, as @bdach mentioned [above](https://github.com/ppy/osu/pull/13349#issuecomment-855348929), the current localisation system is not 100% ready for consumption, so it would better to just hold that off for now.

> @frenzibyte Based on the scale of your requested changes, should I close/draft this PR and come back with a new one?

Probably don't need if you'll stay working on this.

ribbanya commented:
> in `SkinSelectionHandler` itself, and it would be updated whenever any of the menu items are clicked. Though now that I think about it, could just go with the latter proposal anyway (`SkinAnchor` enum that has a `Closest` as well) and use that for the Anchor menu item, would be simpler than the bindables way.

Are you saying that instead of having an anchor setting memorized for each `Drawable`, there's just one setting for the entire `SkinSelectionHandler`? And it defaults to `Closest`? Should that setting be invalidated when the selection is changed in any way (I guess by overriding `HandleSelected`)?

Because [currently](https://github.com/ppy/osu/blob/cce02200601332c1a3bda41cbfc483e93ffafe6c/osu.Game/Skinning/Editor/SkinSelectionHandler.cs#L206), the state is stored in each `Drawable`, which is why I had to use a table to maintain that.

frenzibyte commented:
Ah right, there can be more than one drawable in the selection handler. In that case, you can have that `ConditionalWeakTable` existing with the value type being `SkinAnchor` instead. (and with more shortened names, like `skinAnchors` for example)

Though that brings up the question, how do you imagine "closest" anchors being saved when opening the skin editor later on? sounds like it should be serialised as part of the `SkinnableInfo` but I don't think that can be easily done yet.

peppy commented on osu.Game/Localisation/SkinEditorStrings.cs:
```diff
@@ -0,0 +1,74 @@
+
+namespace osu.Game.Localisation
+{
+    public static class SkinEditorStrings
```
Remove from this PR; out of scope.

peppy commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -4,22 +4,152 @@
+        /// orphaned references to a Drawable forever, unless manually pruned</p>
+        /// <p><see cref="BindableBool"/> is used as a thin wrapper around <see cref="Boolean">bool</see> because ConditionalWeakTable requires a reference type as both a key and a value.</p>
+        /// </remarks>
+        private readonly ConditionalWeakTable<Drawable, BindableBool> isDrawableUsingClosestAnchorLookup = new ConditionalWeakTable<Drawable, BindableBool>();
```
This will need to be stored within `SkinnableInfo` so it can be serialised (as mentioned elsewhere). Just adding a `bool` to track it is fine.

ribbanya replied:
Does this mean you prefer `Anchor` + `bool` over `SkinAnchor` throughout?

peppy replied:
Yeah, I think that will result in better code.

frenzibyte replied:
Right, the calculated anchor will still have to be stored so that drawables receive their actual anchors even when closest is set, I agree with the `Anchor` + `bool` then.

ribbanya commented:
> Though that brings up the question, how do you imagine "closest" anchors being saved when opening the skin editor later on? sounds like it should be serialised as part of the `SkinnableInfo` but I don't think that can be easily done yet.

Yeah, I guess that would have to go in `SkinnableInfo`. Are you saying it can't be done because it would break existing data?

ribbanya committed:
Replace localised strings with static English

ribbanya committed:
Move lookup logic to DrawableExtensions

This is now a global lookup to be shared by serialization and editor.

ribbanya committed:
Add serialization support via SkinnableInfo

gagahpangeran commented on osu.Game/Screens/Play/HUD/SkinnableInfo.cs:
```diff
@@ -32,6 +32,13 @@ public class SkinnableInfo : IJsonSerializable
+        /// automatically determined by proximity, <see langword="true"/> if the user has overridden it.
+        /// </summary>
+        /// <remarks>Stored this way because <c>default(bool)</c> is <see langword="false"/> and we want the default behaviour to be "closest".</remarks>
+        public bool IsNotUsingClosestAnchor { get; set; }
```
```suggestion
        public bool IsUsingClosestAnchor { get; set; }
```

Generally, it is good practice to use positive term for boolean property/variable. Also I see so many negation when using this property in this PR, so just change it to avoid that.

ribbanya replied:
I just wanted it to be very clear about what it's storing. Would something like `OverrideAnchor` be better?

As I mention in the doc comment, it must be stored in a way that `false` means "use closest."

frenzibyte replied:
Could append `= true` to the line in that case? also regarding name, I would shorten it to `UsingClosestAnchor`.

ribbanya replied:
> Could append `= true` to the line in that case? also regarding name, I would shorten it to `UsingClosestAnchor`.

I'm not sure if that works with the serialization, but that would be preferable from a code quality perspective. I'll test it out.

gagahpangeran commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -20,6 +20,85 @@ namespace osu.Game.Skinning.Editor
+
+            var result = default(Anchor);
+
+            static Anchor getTieredComponent(float component, Anchor tier0, Anchor tier1, Anchor tier2) =>
+                component >= 2 / 3f
+                    ? tier2
+                    : component >= 1 / 3f
+                        ? tier1
+                        : tier0;
```
Please don't nested ternary condition. This can be a single method itself.

gagahpangeran reviewed (changes requested):
As mentioned. I just do skimming for now.

ribbanya committed:
Refactor getTieredComponent

ribbanya committed:
Uninvert logic of SkinnableInfo.UsingClosestAnchor

Also rename "IsUsingClosestAnchor" to simply "UsingClosestAnchor".

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -20,6 +20,89 @@ namespace osu.Game.Skinning.Editor
+        /// <summary>
+        /// The hash code of the "Closest" menu item in the anchor point context menu.
+        /// </summary>
+        /// <remarks>Needs only be constant and distinct from any <see cref="Anchor"/> values.</remarks>
+        private static readonly int closest_text_hash = closest_text.GetHashCode();
```
There is nothing actually guaranteeing that `closest_text.GetHashCode()` doesn't accidentally return a value that clashes with the `Anchor` enum values, except for `GetHashCode()` implementation details - which should not be relied upon. I think that all of this hashing is honestly overdone, and an actual simpler method to do this would be to declare an `enum` (just for creating/displaying the items) or another explicitly-defined mapping.

frenzibyte replied:
I was experimenting with this before your review, and I can see this being done with a separate menu item instead, pretty much shortened and easier to read:
```diff
diff --git a/osu.Game/Skinning/Editor/SkinSelectionHandler.cs b/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
index f8d41d94fa..783ed6e27f 100644
--- a/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
+++ b/osu.Game/Skinning/Editor/SkinSelectionHandler.cs
@@ -20,41 +20,6 @@ namespace osu.Game.Skinning.Editor
 {
     public class SkinSelectionHandler : SelectionHandler<ISkinnableDrawable>
     {
-        private const string closest_text = @"Closest";
-
-        /// <summary>
-        /// The hash code of the "Closest" menu item in the anchor point context menu.
-        /// </summary>
-        /// <remarks>Needs only be constant and distinct from any <see cref="Anchor"/> values.</remarks>
-        private static readonly int closest_text_hash = closest_text.GetHashCode();
-
-        private static readonly Dictionary<int, string> anchor_menu_items;
-        private static readonly Dictionary<int, string> origin_menu_items;
-
-        static SkinSelectionHandler()
-        {
-            var anchorMenuSubset = new[]
-            {
-                Anchor.TopLeft,
-                Anchor.TopCentre,
-                Anchor.TopRight,
-                Anchor.CentreLeft,
-                Anchor.Centre,
-                Anchor.CentreRight,
-                Anchor.BottomLeft,
-                Anchor.BottomCentre,
-                Anchor.BottomRight,
-            };
-
-            var texts = anchorMenuSubset.Select(a => a.ToString()).Prepend(closest_text).ToArray();
-            var hashes = anchorMenuSubset.Cast<int>().Prepend(closest_text_hash).ToArray();
-
-            var anchorPairs = hashes.Zip(texts, (k, v) => new KeyValuePair<int, string>(k, v)).ToArray();
-            anchor_menu_items = new Dictionary<int, string>(anchorPairs);
-            var originPairs = anchorPairs.Where(pair => pair.Key != closest_text_hash);
-            origin_menu_items = new Dictionary<int, string>(originPairs);
-        }
-
         private Anchor getClosestAnchorForDrawable(Drawable drawable)
         {
             var parent = drawable.Parent;
@@ -267,35 +232,49 @@ protected override void OnSelectionChanged()

         protected override IEnumerable<MenuItem> GetContextMenuItemsForSelection(IEnumerable<SelectionBlueprint<ISkinnableDrawable>> selection)
         {
-            int checkAnchor(Drawable drawable) =>
-                drawable.UsingClosestAnchor().Value
-                    ? closest_text_hash
-                    : (int)drawable.Anchor;
-
             yield return new OsuMenuItem(nameof(Anchor))
             {
-                Items = createAnchorItems(anchor_menu_items, checkAnchor, applyAnchor).ToArray()
+                Items = new[]
+                {
+                    new TernaryStateRadioMenuItem("Closest", MenuItemType.Standard, _ => applyClosestAnchor())
+                    {
+                        State = { Value = GetStateFromSelection(selection, c => ((Drawable)c.Item).UsingClosestAnchor().Value) }
+                    },
+                }.Concat(createAnchorItems((d, a) => d.Anchor == a && !d.UsingClosestAnchor().Value, applyCustomAnchor)).ToList()
             };

             yield return new OsuMenuItem(nameof(Origin))
             {
                 // Origins can't be "closest" so we just cast to int
-                Items = createAnchorItems(origin_menu_items, d => (int)d.Origin, applyOrigin).ToArray()
+                Items = createAnchorItems((d, o) => d.Origin == o, applyOrigin).ToArray()
             };

             foreach (var item in base.GetContextMenuItemsForSelection(selection))
                 yield return item;

-            IEnumerable<TernaryStateMenuItem> createAnchorItems(IDictionary<int, string> items, Func<Drawable, int> checkFunction, Action<int> applyFunction
) =>
-                items.Select(pair =>
+            IEnumerable<TernaryStateMenuItem> createAnchorItems(Func<Drawable, Anchor, bool> checkFunction, Action<Anchor> applyFunction)
+            {
+                var displayableAnchors = new[]
                 {
-                    var (hash, text) = pair;
-
-                    return new TernaryStateRadioMenuItem(text, MenuItemType.Standard, _ => applyFunction(hash))
+                    Anchor.TopLeft,
+                    Anchor.TopCentre,
+                    Anchor.TopRight,
+                    Anchor.CentreLeft,
+                    Anchor.Centre,
+                    Anchor.CentreRight,
+                    Anchor.BottomLeft,
+                    Anchor.BottomCentre,
+                    Anchor.BottomRight,
+                };
+
+                return displayableAnchors.Select(a =>
+                {
+                    return new TernaryStateRadioMenuItem(a.ToString(), MenuItemType.Standard, _ => applyFunction(a))
                     {
-                        State = { Value = GetStateFromSelection(selection, c => checkFunction((Drawable)c.Item) == hash) }
+                        State = { Value = GetStateFromSelection(selection, c => checkFunction((Drawable)c.Item, a)) }
                     };
                 });
+            }
         }

         private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpacePosition)
@@ -304,19 +283,15 @@ private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpac
                 drawable.Parent.ToLocalSpace(screenSpacePosition) - drawable.AnchorPosition;
         }

-        private void applyOrigin(int hash)
+        private void applyOrigin(Anchor origin)
         {
-            var anchor = (Anchor)hash;
-
             foreach (var item in SelectedItems)
             {
                 var drawable = (Drawable)item;

                 var previousOrigin = drawable.OriginPosition;
-                drawable.Origin = anchor;
+                drawable.Origin = origin;
                 drawable.Position += drawable.OriginPosition - previousOrigin;
-
-                updateDrawableAnchorIfUsingClosest(drawable);
             }
         }

@@ -327,39 +302,35 @@ private void applyOrigin(int hash)
         private Quad getSelectionQuad() =>
             GetSurroundingQuad(SelectedBlueprints.SelectMany(b => b.Item.ScreenSpaceDrawQuad.GetVertices().ToArray()));

-        private void applyAnchor(int hash)
+        private void applyClosestAnchor()
         {
             foreach (var item in SelectedItems)
             {
                 var drawable = (Drawable)item;

-                var anchor = getAnchorFromHashAndDrawableAndRecordWhetherUsingClosestAnchor(hash, drawable);
+                drawable.UsingClosestAnchor().Value = true;
+                updateDrawableAnchor(drawable, getClosestAnchorForDrawable(drawable));
+            }
+        }
+
+        private void applyCustomAnchor(Anchor anchor)
+        {
+            foreach (var item in SelectedItems)
+            {
+                var drawable = (Drawable)item;

+                drawable.UsingClosestAnchor().Value = false;
                 updateDrawableAnchor(drawable, anchor);
             }
         }

-        private static void updateDrawableAnchor(Drawable drawable, Anchor anchor)
+        private void updateDrawableAnchor(Drawable drawable, Anchor anchor)
         {
             var previousAnchor = drawable.AnchorPosition;
             drawable.Anchor = anchor;
             drawable.Position -= drawable.AnchorPosition - previousAnchor;
         }

-        private Anchor getAnchorFromHashAndDrawableAndRecordWhetherUsingClosestAnchor(int hash, Drawable drawable)
-        {
-            var isUsingClosestAnchor = drawable.UsingClosestAnchor();
-
-            if (hash == closest_text_hash)
-            {
-                isUsingClosestAnchor.Value = true;
-                return getClosestAnchorForDrawable(drawable);
-            }
-
-            isUsingClosestAnchor.Value = false;
-            return (Anchor)hash;
-        }
-
         private static void adjustScaleFromAnchor(ref Vector2 scale, Anchor reference)
         {
             // cancel out scale in axes we don't care about (based on which drag handle was used).
```

ribbanya replied:
> There is nothing actually guaranteeing that `closest_text.GetHashCode()` doesn't accidentally return a value that clashes with the `Anchor` enum values, except for `GetHashCode()` implementation details - which should not be relied upon.

This can be solved with a static computation (something simple like `~(Anchor.x0 | Anchor.x1 | Anchor.x2 | Anchor.y0 | Anchor.y1 | Anchor.y2)` would meet the requirements).

> I think that all of this hashing is honestly overdone, and an actual simpler method to do this would be to declare an `enum` (just for creating/displaying the items) or another explicitly-defined mapping.

So... I tried this, and it's a ton of boilerplate and more error-prone. I think one hash plus an existing `enum` is more elegant, but let me know what you think.

I could finish a working implementation but I think this is enough of an illustration:
```cs
private enum MenuAnchor
{
    Closest,
    TopLeft,
    TopCentre,
    TopRight,
    CentreLeft,
    Centre,
    CentreRight,
    BottomLeft,
    BottomCentre,
    BottomRight
}

private static MenuAnchor getMenuAnchorForDrawable(Drawable drawable)
{
    if (drawable.UsingClosestAnchor().Value)
        return MenuAnchor.Closest;

    const Anchor mask = Anchor.x0 | Anchor.x1 | Anchor.x2 | Anchor.y0 | Anchor.y1 | Anchor.y2;

    switch (drawable.Anchor & mask)
    {
        case Anchor.TopLeft:
            return MenuAnchor.TopLeft;

        case Anchor.TopCentre:
            return MenuAnchor.TopCentre;

        case Anchor.TopRight:
            return MenuAnchor.TopRight;

        case Anchor.CentreLeft:
            return MenuAnchor.CentreLeft;

        case Anchor.Centre:
            return MenuAnchor.Centre;

        case Anchor.CentreRight:
            return MenuAnchor.CentreRight;

        case Anchor.BottomLeft:
            return MenuAnchor.BottomLeft;

        case Anchor.BottomCentre:
            return MenuAnchor.BottomCentre;

        case Anchor.BottomRight:
            return MenuAnchor.BottomRight;

        default:
            throw new ArgumentOutOfRangeException(nameof(drawable), @"Drawable.Anchor is invalid.");
    }
}
```

An alternative implementation would be one where `MenuAnchor` copies the flags from `Anchor`, but at that point it's just a cast anyway.

ribbanya replied:
> I was experimenting with this before your review, and I can see this being done with a separate menu item instead, pretty much shortened and easier to read

I have this saved as a stash but I want to resolve the `enum` question first. I think the approach is fundamentally more sound than what I had, though.

bdach replied:
If I'm honest, I prefer boilerplate to unnecessary complexity, and the hashing stuff firmly places in the second camp to me. I haven't read @frenzibyte's suggestion overly thoroughly but it seemed more tolerable for sure.

ribbanya replied:
Okay, I'll proceed in that direction.

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -20,6 +20,89 @@ namespace osu.Game.Skinning.Editor
+            if (parent == null)
+                return drawable.Anchor;
+
+            // If there is a better way to get this information, let me know. Was taken from LogoTrackingContainer.ComputeLogoTrackingPosition
+            // I tried a lot of different things, such as just using Position / ChildSize, but none of them worked properly.
```
This comment needs to disappear before this is merged - first-person comments are not the right way to do this. For this type of commentary, you can use self-posted review comments on the actual github PR.

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -234,18 +327,39 @@ private void applyOrigin(Anchor anchor)
+            drawable.Position -= drawable.AnchorPosition - previousAnchor;
+        }
+
+        private Anchor getAnchorFromHashAndDrawableAndRecordWhetherUsingClosestAnchor(int hash, Drawable drawable)
```
This method name is too much. Something like `mapHashToAnchorSettings` tells pretty much the same story. For expanded commentary, comments/xmldoc can be used, if necessary.

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -20,6 +20,89 @@ namespace osu.Game.Skinning.Editor
+
+            // If there is a better way to get this information, let me know. Was taken from LogoTrackingContainer.ComputeLogoTrackingPosition
+            // I tried a lot of different things, such as just using Position / ChildSize, but none of them worked properly.
+            var screenPosition = getOriginPositionFromQuad(drawable.ScreenSpaceDrawQuad, drawable.Origin);
```
I *believe* there's a chance you could multiply the quad size by `RelativeOriginPosition` and get the same result, but it's not a huge deal which way is actually used here.

ribbanya committed:
Remove first-person comment

ribbanya committed:
Rename overly long method

ribbanya commented on osu.Game/Extensions/DrawableExtensions.cs:
```diff
@@ -56,12 +60,34 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
+        /// <a href="https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.conditionalweaktable-2?view=net-5.0#thread-safety">documentation</a>,
+        /// but the <c>BindableBool</c>s are not unless <see cref="Bindable{T}.BeginLease">leased</see>.</p>
+        /// </remarks>
+        private static readonly ConditionalWeakTable<IDrawable, BindableBool> is_drawable_using_closest_anchor_lookup = new ConditionalWeakTable<IDrawable, BindableBool>();
```
> This cannot exist. Game-global dictionaries storing drawable references are a huge no-no. 

Makes sense.

> This needs to be localised more to the skin editor itself in some capacity.

Is this just a matter of finding where to put a `[Cached]` storage wrapper around the `ConditionalWeakTable`?

ribbanya replied:
Also, I'm not saying that `DrawableExtensions` is a good place for this—it's not, and I will change this to be osu-compliant. I understand and agree with you in principal that holding onto global references is bad. But I'm also having trouble seeing the pitfalls that you're suggesting are present in this specific implementation.

In its current form, `UsingClosestAnchor()` will never give back a reference to a `Drawable`; it only takes references in. If a key (a `Drawable`) in the table is not referenced externally, the entry is deleted by `ConditionalWeakTable` and garbage-collected. A `BindableBool` has no concept of a `Drawable`, so no references to it are held or exposed in that way.

So the worst-case scenario seems to be that two separate classes, which independently already have a reference to a given `Drawable`, are able to obtain a common `BindableBool` associated with it, but that `BindableBool` has no references to or from the original `Drawable`. So, as I pointed out in the xmldoc, there could be threading issues when writing to the underlying value of the shared `BindableBool`.

I'm only bringing this up so that I'm clear on the drawbacks in upcoming iterations.

bdach replied:
Caching the collection somewhere in the skin editor hierarchy would be the obvious way to proceed, yes.

As to the implementation as it reads currently: I did not mean to suggest (or even want to argue about) the correctness of the reference-holding logic. The reason I rejected it so sternly is quite a bit simpler than that:

1. This project has a hard-line stance against shared global state, and provides mutliple facilities and patterns designed to avoid keeping such state or helping manage it if it really does need to exist. I don't think I need to explain why shared global state is bad (very hard to debug, leads to "spooky action at a distance"). In this case it is not necessary to store global state, so it should not be done.
2. Having an extension method *store* global state is extremely non-idiomatic. I have been reading C# code for quite a few years now, and have never seen anything of the sort. What extension methods are usually intended for is usually small pieces of convenience code to extend a class that cannot be extended in another manner.

ribbanya replied:
That makes a lot of sense. Thanks for being patient with me.

ribbanya commented:
Would it be better to add
```cs
UsingClosestAnchor { get; set; }
```

to `ISkinnableDrawable` instead of this duck-typing with a table? If we do that, it would make more sense to invert the logic everywhere (`UsingClosestAnchor` = `true` → `OverridesClosestAnchor` = `false`), I think.

Currently would affect these classes:

1. `LegacyHealthDisplay`
1. `LegacyAccuracyCounter`
1. `LegacyScoreCounter`
1. `SkinnableTargetComponentsContainer`
1. `DefaultAccuracyCounter`
1. `DefaultComboCounter`
1. `DefaultHealthDisplay`
1. `DefaultScoreCounter`
1. `HitErrorMeter`
1. `LegacyComboCounter`
1. `SongProgress`

peppy commented:
Without having read through your full implementation, that sounds like the initial direction I would have tried, since the implementation complexity should be far lower.

ribbanya commented:
> Without having read through your full implementation, that sounds like the initial direction I would have tried, since the implementation complexity should be far lower.

Yeah, it's what I would have done if it were my own codebase. I just did something silly instead because I was afraid of modifying too many files.

I'll come back soon with something more reasonable.

ribbanya committed:
Add UsingClosestAnchor to ISkinnableDrawable

Also implement it as an auto property in its inheritors.
The auto properties default to true.

ribbanya committed:
Remove all UsingClosestAnchor() extension logic

It is replaced with ISkinnableDrawable.UsingClosestAnchor.

ribbanya committed:
Rename UsingClosestAnchor

It is now "OverridesClosestAnchor". The logic is inverted accordingly.

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -234,18 +325,29 @@ private void applyOrigin(Anchor anchor)
+                var (overridesClosest, anchor) =
+                    hash == closest_text_hash
+                        ? (false, getClosestAnchorForDrawable(drawable))
+                        : (true, (Anchor)hash);
```
Is this syntax too cute?

peppy replied:
This is nigh impossible to read.

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -151,11 +232,24 @@ public override bool HandleMovement(MoveSelectionEvent<ISkinnableDrawable> moveE
 
+        private void updateDrawableAnchorIfUsingClosest(Drawable drawable)
+        {
+            if (drawable is ISkinnableDrawable { OverridesClosestAnchor: true }) return;
```
Is pattern matching like this acceptable?

peppy replied:
No.

ribbanya committed:
Refactor unacceptable syntax

frenzibyte commented on osu.Game/Skinning/ISkinnableDrawable.cs:
```diff
@@ -14,5 +14,11 @@ public interface ISkinnableDrawable : IDrawable
+        /// <see langword="false"/> if this <see cref="ISkinnableDrawable"/>'s <see cref="Drawable.Anchor"/> is automatically determined by proximity,
+        /// <see langword="true"/> if the user has overridden it.
+        /// </summary>
+        bool OverridesClosestAnchor { get; set; }
```
I think you could get away with default interface implementation instead, that would work amicably without having to implement this everywhere:
```suggestion
        public bool OverridesClosestAnchor { get; set; }
```

frenzibyte replied:
Also not really sure about the change from `UsingClosestAnchor` to `OverridesClosestAnchor`, the old way reads better IMO. (could even rename to `UsesClosestAnchor`)

ribbanya replied:
Do you mean the C# 8 feature? I tried that but I don't believe those can hold any kind of state (auto properties create implicit fields).

bdach replied:
Unresolving this because I don't like `OverridesClosestAnchor` either. `UsesClosestAnchor` seems better. Or if you want to go the converse route, `UsesFixedAnchor` as a negation of that.

ribbanya committed:
Simplify and rearrange SkinSelectionHandler

The file has been restructured and reworded such that there are as few
differences as possible from b36b40cb3430e6bd511390f82c39ba23195dc8cb.

ribbanya commented:
Okay, everything is a lot more sane now.

Sorry for all the missteps; I definitely learned a lot from this PR. Thank you to everyone who gave me advice.

frenzibyte commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -333,22 +255,75 @@ private void applyOrigin(int hash)
             GetSurroundingQuad(SelectedBlueprints.SelectMany(b => b.Item.ScreenSpaceDrawQuad.GetVertices().ToArray()));
 
-        private void applyAnchor(int hash)
+        private void applyAnchor(Anchor anchor)
```
I will suggest renaming this to `applyCustomAnchor` or something, given that this turns on `OverridesClosestAnchor`.

ribbanya committed:
Rename `applyAnchor` to `applyCustomAnchor`

bdach commented on osu.Game/Extensions/DrawableExtensions.cs:
```diff
@@ -57,6 +58,9 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
             component.Anchor = info.Anchor;
             component.Origin = info.Origin;
 
+            if (component is ISkinnableDrawable skinnable)
+                skinnable.OverridesClosestAnchor = info.OverridesClosestAnchor;
```
Not super keen on this type check, especially combined with the TODO above. Given that now `ISkinnableDrawable` is becoming more specific to the skinning setup than `Drawable`, I think a good way forward here would be to add the other values being applied in this method (ie. `Position`, `Rotation`, etc.) to `ISkinnableDrawable`. That way `SkinnableInfo` could also implement that interface, and there could be a general-purpose
```csharp
public static void CopyFrom(this ISkinnableInfo target, ISkinnableInfo source)
```
method that can be reused both here and in the `SkinnableInfo` ctor. Not super sure whether that should happen as part of this PR or not, though.

ribbanya replied:
The type check is definitely ugly. Do you want me to open a new PR that lessens the usage of `Drawable` in favor of `ISkinnableDrawable`?

bdach replied:
Not 100% sure myself. Would like input from others as to whether that sounds like a logical direction to take.

frenzibyte replied:
I believe `SkinnableInfo` wouldn't even have to exist and the drawable would be serialised/populated using the `ISkinnableDrawable` interface? that I would agree to.

bdach replied:
Not entirely sure serialising interfaces is a thing that will work super well. May still need `SkinnableInfo` as a POCO intermediary if JSON libs will not comply.

ribbanya replied:
Yeah, [this](https://www.c-sharpcorner.com/UploadFile/20c06b/deserializing-interface-properties-with-json-net/) looks like a mess to me compared to just a POCO.

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -151,11 +151,26 @@ public override bool HandleMovement(MoveSelectionEvent<ISkinnableDrawable> moveE
             return true;
         }
 
+        private void updateDrawableAnchorIfUsingClosest(ISkinnableDrawable item)
```
The way these anchor application methods are written is really confusing to me. There are several code paths that do slightly different things. This and `applyCustomAnchor` should really be one method.

bdach replied:
I think `UsesFixedAnchor` is also fine, fwiw. As long as `Overrides` isn't in the name, since that naming is also a bit backwards (as it specifies which behaviour *is turned off*, and not which one *is actually used*).

ribbanya replied:
Ah, okay. I was conceiving the `Overrides` as "a thing the user did to the object." I'll go with `UsesFixedAnchor`, then.

ribbanya committed:
Rename `OverridesClosestAnchor` to `UsesFixedAnchor`

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -234,18 +255,82 @@ private void applyOrigin(Anchor anchor)
+                item.OverridesClosestAnchor = false;
+                updateDrawableAnchor(drawable, getClosestAnchorForDrawable(drawable));
             }
         }
```
@bdach If I change these to something like this:
```cs
/// <param name="anchor">An <see cref="Anchor"/> value to apply a fixed anchor point,
/// or <see langword="null"/> if the closest point should be used.</param>
applyAnchor(ISkinnableDrawable item, Anchor? anchor)
// ...
```

Then it could be merged with `updateDrawableAnchorIfUsingClosest` as well. The loops could be moved to respective call sites. The problem is that nullable branching parameters are a bit obtuse. Thoughts?

ribbanya replied:
@bdach If these are combined, the iteration logic gets a little weird. Like I could make another method that just calls `applyClosestAnchor` to everything in `SelectedItems`, but that seems a bit weird to me?

Do you think just inlining `updateDrawableAnchorIfUsingClosest` makes more sense? It only occurs once.

```cs
public override bool HandleMovement(MoveSelectionEvent<ISkinnableDrawable> moveEvent)
{
    foreach (var c in SelectedBlueprints) // <- loop is here
    {
        Drawable drawable = (Drawable)c.Item;
        // ... and has to do this:
        drawable.Position += drawable.ScreenSpaceDeltaToParentSpace(moveEvent.ScreenSpaceDelta);

        updateDrawableAnchorIfUsingClosest(c.Item);
    }

    return true;
}

private void applyClosestAnchor()
{
    foreach (var item in SelectedItems) // <- loop again, but just does the anchor stuff
    {
        var drawable = (Drawable)item;

        item.UsesFixedAnchor = false;
        updateDrawableAnchor(drawable, getClosestAnchorForDrawable(drawable));
    }
}

private void updateDrawableAnchorIfUsingClosest(ISkinnableDrawable item)
{
    // these conditionals are unique to HandleMovement:
    if (item.UsesFixedAnchor) return;

    var drawable = (Drawable)item;

    var closestAnchor = getClosestAnchorForDrawable(drawable);

    if (closestAnchor == drawable.Anchor) return;

    // the common code is already a function:
    updateDrawableAnchor(drawable, closestAnchor);
}
```

ribbanya replied:
Actually if I just delete the `if (closestAnchor == drawable.Anchor) return;` then the code can be merged. But the looping problem is still there.

ribbanya committed:
Inline updateDrawableAnchorIfUsingClosest

ribbanya committed:
Refactor `updateDrawableAnchorIfUsingClosest`

ribbanya committed:
Rename some methods for clarity

Methods which operate on a collection of `ISkinnableDrawable`s are now
plural; ones which take a single item are singular.

This also allows cutting down the name of `getClosestAnchorForDrawable`
to just `getClosestAnchor`.

ribbanya committed:
Move guard clause from `checkAndApplyClosestAnchor` to `applyAnchor`

ribbanya committed:
Simplify `applyClosestAnchor` to one line by moving another guard clause

ribbanya committed:
Rename `applyCustomAnchors` to `applyFixedAnchors` for consistency with `UsesFixedAnchor`

ribbanya committed:
Add guard clause to `applyOrigins` and rename parameter

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -215,14 +229,16 @@ private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpac
             {
                 var drawable = (Drawable)item;
 
+                if (origin == drawable.Origin) continue;
```
I don't know how necessary this and...
https://github.com/ppy/osu/blob/10b6b7290923a112dfee58c8a8c820dc32934611/osu.Game/Skinning/Editor/SkinSelectionHandler.cs#L322
... are, but I wanted them to at least be consistent. In the case that some drawables already have the correct anchor/origin, it's possible that by moving them slightly it'll cause some jitter. Probably sub-pixel, though.

peppy replied:
I'm not sure where the "pixel" or "subpixel" part comes from. We are using floats, which are never going to be "pixel" based, nor are they ever round to `int` values or anything.

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -234,18 +250,82 @@ private void applyOrigin(Anchor anchor)
+
+            var screenPosition = getOriginPositionFromQuad(drawable.ScreenSpaceDrawQuad, drawable.Origin);
+            var absolutePosition = parent.ToLocalSpace(screenPosition);
+            var factor = parent.RelativeToAbsoluteFactor;
+
+            var result = default(Anchor);
+
+            result |= getTieredComponent(absolutePosition.X / factor.X, Anchor.x0, Anchor.x1, Anchor.x2);
+            result |= getTieredComponent(absolutePosition.Y / factor.Y, Anchor.y0, Anchor.y1, Anchor.y2);
+
+            return result;
+        }
+
+        private static Vector2 getOriginPositionFromQuad(in Quad quad, Anchor origin)
+        {
+            var result = quad.TopLeft;
+
+            if (origin.HasFlagFast(Anchor.x2))
+                result.X += quad.Width;
+            else if (origin.HasFlagFast(Anchor.x1))
+                result.X += quad.Width / 2f;
+
+            if (origin.HasFlagFast(Anchor.y2))
+                result.Y += quad.Height;
+            else if (origin.HasFlagFast(Anchor.y1))
+                result.Y += quad.Height / 2f;
+
+            return result;
+        }
+
+        private static Anchor getTieredComponent(float component, Anchor tier0, Anchor tier1, Anchor tier2)
+        {
+            if (component >= 2 / 3f)
+                return tier2;
+
+            if (component >= 1 / 3f)
+                return tier1;
+
+            return tier0;
+        }
```
Should this just be moved to `DrawableExtensions` (and `getClosestAnchor` be made `public`)? Nothing else is using it, but it would reduce some bloat of `SkinSelectionHandler`, and it is a pure function.

peppy replied:
make it a local function inside the only usage site. also don't use the word "tier". probably just `anchor0` .. `anchor2`.

peppy commented on osu.Game/Screens/Play/HUD/SkinnableInfo.cs:
```diff
@@ -32,6 +32,9 @@ public class SkinnableInfo : IJsonSerializable
 
         public Anchor Origin { get; set; }
 
+        /// <inheritdoc cref="ISkinnableDrawable.UsesFixedAnchor"/>
```
We generally don't use `inheritdoc`. It's automatically handled in IDEs these days without its presence.

peppy replied:
Okay so this isn't actually inheriting from a direct interface. Just add local documentation here then, probably referencing how it is consumed/sourced from the related property on the interface.

peppy replied:
This documentation should mention that this is only used in the skin layout editor context.

I'd almost be inclined to invert the naming to `UseClosestAnchor` too..

ribbanya replied:
> I'd almost be inclined to invert the naming to `UseClosestAnchor` too..

This commit (on a [separate branch](https://github.com/ribbanya/osu/tree/skin-editor-closest-anchor-invert)) will do that: https://github.com/ppy/osu/commit/7dc8f9aafa165d36c7604dd7b70223ea5970bb9d

The serialization does work, I just think the default-true semantics are a little wonky when implementing the interface. Up to you.

peppy replied:
That branch presents an issue: the default implementations should not be specifying `true` for this property. It feels completely weird.

ribbanya replied:
Yeah, exactly. I'm not aware of a better way to invert the property. Will implement if anyone has any suggestions. Otherwise, I guess it will have to be `UsesFixedAnchor`.

The fundamental problem is that when an `ISkinnableDrawable` which has never been serialized is encountered by the editor, the only information it has is the CLR default value for the property. Which means that if we want "closest anchor" behaviour to be the default, and `UsesClosestAnchor` is the name of the property, that property must be `true` when a new element is instantiated.

peppy replied:
I think it's okay with the behaviour inverted. I'm hoping we can find a better way of defining this skinnable interface in a future, but it should be fine for now to get this functionality actually in the game.

peppy commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -215,14 +229,16 @@ private static void updateDrawablePosition(Drawable drawable, Vector2 screenSpac
         }
 
-        private void applyOrigin(Anchor anchor)
+        private void applyOrigins(Anchor origin)
```
Why were these methods renamed?

ribbanya replied:
> Why were these methods renamed?

There were two components to the rename. One is that `applyAnchor` was split into `applyFixedAnchor` and `applyClosestAnchor`, the latter of which didn't exist before.

The second is that member functions which loop over the selection (`applyFixedAnchors`, `applyClosestAnchors`, `applyOrigins`) were changed to be plural, and static functions that operate on a single `Drawable` (`getClosestAnchor`, `applyAnchor`) were made singular.

peppy commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -171,20 +179,27 @@ protected override void OnSelectionChanged()
             yield return new OsuMenuItem("Origin")
             {
-                Items = createAnchorItems(d => d.Origin, applyOrigin).ToArray()
+                Items = createAnchorItems((i, o) => ((Drawable)i).Origin == o, applyOrigins).ToArray()
```
```suggestion
                Items = createAnchorItems((d, o) => ((Drawable)d).Origin == o, applyOrigins).ToArray()
```

peppy commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -171,20 +179,27 @@ protected override void OnSelectionChanged()
             yield return new OsuMenuItem("Anchor")
             {
-                Items = createAnchorItems(d => d.Anchor, applyAnchor).ToArray()
+                Items = createAnchorItems((i, a) => i.UsesFixedAnchor && ((Drawable)i).Anchor == a, applyFixedAnchors)
```
```suggestion
                Items = createAnchorItems((d, a) => d.UsesFixedAnchor && ((Drawable)d).Anchor == a, applyFixedAnchors)
```

peppy commented:
The closest anchor should probably update when the origin is changed by the user:

https://user-images.githubusercontent.com/191335/121643638-00797b80-cacd-11eb-9968-c3ee168f657c.mp4

From a usability perspective, I find it weird that when dragging a selection the "closest" anchor is not the same for all items in the selection. But this might be fine for now, as there are likely cases the user would actually want this behaviour (ie. if they are moving all the elements from the top of their screen to the bottom, and still expecting them to anchor left/middle/right). Hard to say.

https://user-images.githubusercontent.com/191335/121643927-58b07d80-cacd-11eb-9234-8df33e1a10d4.mp4

peppy commented on osu.Game/Skinning/ISkinnableDrawable.cs:
```diff
@@ -14,5 +14,11 @@ public interface ISkinnableDrawable : IDrawable
         bool IsEditable => true;
+
+        /// <summary>
+        /// <see langword="false"/> if this <see cref="ISkinnableDrawable"/>'s <see cref="Drawable.Anchor"/> is automatically determined by proximity,
```
This also needs to mention that it's *only in the context of editing the layout*. Right now it sounds like it's always going to be determined on application, which is not correct.

ribbanya committed:
Revert rename of lambda variables

ribbanya committed:
Recalculate closest anchor when origin is changed

ribbanya committed:
Make `getTieredComponent` local

ribbanya committed:
Rename parameters of `getTieredComponent`

ribbanya committed:
Rename `getTieredComponent` to `getAnchorFromPosition`

Also rename parameter `component` to `xOrY`.

ribbanya committed:
Make `getOriginPositionFromQuad` local

ribbanya committed:
Inline `getOriginPositionFromQuad`

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -281,24 +281,21 @@ private static Anchor getClosestAnchor(Drawable drawable)
+                    screenPosition.Y += drawable.ScreenSpaceDrawQuad.Height / 2f;
             }
 
-            var screenPosition = getOriginPositionFromQuad(drawable.ScreenSpaceDrawQuad, drawable.Origin);
```
Which is better?

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -287,19 +287,19 @@ private static Anchor getClosestAnchor(Drawable drawable)
+            result |= getAnchorFromPosition(absolutePosition.Y / factor.Y, Anchor.y0, Anchor.y1, Anchor.y2);
 
             return result;
         }
```
Should I keep this?

ribbanya commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -281,24 +281,21 @@ private static Anchor getClosestAnchor(Drawable drawable)
 
-            static Vector2 getOriginPositionFromQuad(in Quad quad, Anchor origin)
+            var screenPosition = drawable.ScreenSpaceDrawQuad.TopLeft;
             {
```
Do we use scope braces here?

bdach replied:
Scope braces are rarely used around these parts. You're probably better off splitting a (local?) method.

ribbanya committed:
Extract `drawable.ScreenSpaceDrawQuad` to a variable

ribbanya commented:
> The closest anchor should probably update when the origin is changed by the user

Fixed in 635300b3115f721188ca3ab6593b5af0cc99c921.

> From a usability perspective, I find it weird that when dragging a selection the "closest" anchor is not the same for all items in the selection.

To do that we'd need to determine the aggregate origin of the selection, which I'm not sure how to do from a design standpoint. Currently origins which are not common to all elements in the selection are `TernaryState.Indeterminate` in the menu.

> But this might be fine for now, as there are likely cases the user would actually want this behaviour (ie. if they are moving all the elements from the top of their screen to the bottom, and still expecting them to anchor left/middle/right). Hard to say.

Could make that another menu item, which would suggest converting the `Anchor`+`bool` approach to a new enum as has been discussed before.

peppy committed:
Merge branch 'master' into skin-editor-closest-anchor

peppy committed:
Reword xmldoc

peppy committed:
Move scoped functionality into local function

peppy committed:
Merge branch 'master' into skin-editor-closest-anchor
